mlkem/kem.c

```diff
@@ -183,34 +180,149 @@ static int mlk_check_pct(uint8_t const pk[MLKEM_INDCCA_PUBLICKEYBYTES],
+{
+  if (mlk_check_sk(sk))
+  {
+    return -1;
+  }
+  mlk_indcpa_parse_sk(&sks->indcpa_sk, sk);
+  mlk_indcpa_parse_pk(&sks->indcpa_pk, sk + MLKEM_INDCPA_SECRETKEYBYTES);
+  memcpy(sks->hpk,
+         sk + MLKEM_INDCPA_SECRETKEYBYTES + MLKEM_INDCPA_PUBLICKEYBYTES,
+         MLKEM_SYMBYTES);
+  memcpy(sks->z,
+         sk + MLKEM_INDCPA_SECRETKEYBYTES + MLKEM_INDCPA_PUBLICKEYBYTES +
+             MLKEM_SYMBYTES,
+         MLKEM_SYMBYTES);
+  return 0;
+}
+
+
+MLK_EXTERNAL_API
+int crypto_kem_keypair_derand_struct(mlk_public_key *pk, mlk_secret_key *sk,
+                                     const uint8_t coins[2 * MLKEM_SYMBYTES])
+{
+  mlk_indcpa_keypair_derand(&sk->indcpa_pk, &sk->indcpa_sk, coins);
+
+  // pre-compute H(pk)
+  uint8_t pks[MLKEM_INDCCA_PUBLICKEYBYTES];
+  mlk_indcpa_marshal_pk(pks, &sk->indcpa_pk);
+  mlk_hash_h(sk->hpk, pks, MLKEM_INDCCA_PUBLICKEYBYTES);
+
+  // copy over indcpa pk and H(pk) to public key
+  // pk is NULL during parsing before decaps as the pk is not needed
+  if (pk != NULL)
+  {
+    memcpy(&pk->indcpa_pk, &sk->indcpa_pk, sizeof(mlk_indcpa_public_key));
+    memcpy(pk->hpk, sk->hpk, MLKEM_SYMBYTES);
+  }
+
+  // Value z for pseudo-random output on reject
+  memcpy(sk->z, coins + MLKEM_SYMBYTES, MLKEM_SYMBYTES);
+
+  // seed to regenerate whole secret key
+  memcpy(sk->seed, coins, MLKEM_SYMBYTES);
 
   /* Declassify public key */
-  MLK_CT_TESTING_DECLASSIFY(pk, MLKEM_INDCCA_PUBLICKEYBYTES);
+  MLK_CT_TESTING_DECLASSIFY(&pk->indcpa_pk, MLKEM_INDCCA_PUBLICKEYBYTES);
 
   /* Pairwise Consistency Test (PCT) (FIPS 140-3 IPG) */
   if (mlk_check_pct(pk, sk))
   {
     return -1;
   }
 
+
+  /* Specification: Partially implements
+   * [FIPS 203, Section 3.3, Destruction of intermediate values] */
+  mlk_zeroize(pks, sizeof(pks));
   return 0;
 }
 
+MLK_EXTERNAL_API
+int crypto_kem_keypair_struct(mlk_public_key *pk, mlk_secret_key *sk)
+{
+  int res;
+  MLK_ALIGN uint8_t coins[2 * MLKEM_SYMBYTES];
+
+  /* Acquire necessary randomness, and mark it as secret. */
+  randombytes(coins, 2 * MLKEM_SYMBYTES);
+  MLK_CT_TESTING_SECRET(coins, sizeof(coins));
+
+  res = crypto_kem_keypair_derand_struct(pk, sk, coins);
+
+
+  /* Specification: Partially implements
+   * [FIPS 203, Section 3.3, Destruction of intermediate values] */
+  mlk_zeroize(coins, sizeof(coins));
+  return res;
+}
+
+
+/* Reference: `crypto_kem_keypair_derand()` in the reference implementation
+ *            - We optionally include PCT which is not present in
+ *              the reference code. */
+MLK_EXTERNAL_API
+int crypto_kem_keypair_derand(uint8_t pk[MLKEM_INDCCA_PUBLICKEYBYTES],
+                              uint8_t sk[MLKEM_INDCCA_SECRETKEYBYTES],
+                              const uint8_t coins[2 * MLKEM_SYMBYTES])
+{
+  int res;
+  mlk_public_key pks;
+  mlk_secret_key sks;
+
+
+  res = crypto_kem_keypair_derand_struct(&pks, &sks, coins);
+
+
+  crypto_kem_marshal_pk(pk, &pks);
+  crypto_kem_marshal_sk(sk, &sks);
+
```

Comment:
Referring to the PCT comment, I think we need to retain the PCT here for the time being.

Reply:
No - If PCT is enabled, _all_ public APIs need to do PCT - so we cannot move it here, otherwise `crypto_kem_keypair_struct` doesn't peform PCT. We can perform serialization in the PCT if needed, but that is _very_ expensive.

Reply:
The current code does not perform the PCT on the key material that is returned to the caller -- it is thus not clear if it is compliant. Please re-introduce the previous form of the PCT.

I don't know how to best introduce a PCT in the internal-format API -- but this is secondary to me at the moment, because the FIPS-compliance of that API is still questionable anyway.

Reply:
Okay - I have implemented what you said - we should be explicit somewhere that PCT is ONLY implemented when using `crypto_kem_keypair`, but not when using any of the other functions.
AWS-LC may want to consider changing this in the medium-term - it's much faster if you can work on the internal formats.

Reply:
> AWS-LC may want to consider changing this in the medium-term - it's much faster if you can work on the internal formats.

Of course it is -- but it's about FIPS compliance. When it's clear that the internal PCT is still enough, I'm confident it would be changed.